---
title: "Regression models for statistical analysis of gene expression data"
format:
  html:
    df-print: paged
---

## Sources

[A guide to creating design matrices for gene expression experiments](https://f1000research.com/articles/9-1444/v1)

[RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR](https://f1000research.com/articles/5-1408/v3)

## Init

```{r}
#| output: false
library(tidyverse)
library(magrittr)
library(broom)
library(marginaleffects)
library(limma)

theme_set(theme_bw())
```

## Common statistical tests are regression models

<https://lindeloev.github.io/tests-as-linear/>

![](images/image-196232366.png)

## Regression models

Regression models propose that $y_i$ is a function of $x_i$ and $\beta$ with $e_i$ representing an additive error term that may stand in for un-modeled determinants of $y_i$ or random noise. The researcher's goal is to estimate the function $f(x_i, \beta)$ that most closely fits the data generation process based on the data/observations:

$$
\begin{gather}
(y_i,x_i) \\
i = 1, \ldots, n
\end{gather}
$$ Model:

$$
y_i = f(x_i, \beta) + e_i
$$

For example (univariate linear regression):

$$
f(x_i, \beta) = \beta_0 + \beta_1 x_i
$$

$$
y_i = \beta_0 + \beta_1 x_i + e_i
$$ $$
e_i \sim \mathcal{N}(\mu = 0, \sigma^2)
$$

$$
y_i \sim \beta_0 + \beta_1 x_i + \mathcal{N}(\mu = 0, \sigma^2)
$$

$$
y_i \sim \mathcal{N}(\mu_i, \sigma^2)
$$

$$
\mu_i = \beta_0 + \beta_1 x_i = E(Y)
$$

![](images/image-158884090.png)

### Deterministic model

$$
y = a + b x
$$

```{r}
set.seed(2023)

a = 2
b = 2
x = seq(2, 12, by = 2)
y = a + b * x

(data = tibble(x, y))
```

```{r}
ggplot(data, aes(x = x, y = y)) +
  geom_point() +
  geom_abline(intercept = a, slope = b)
```

### Stochastic/statistical model

```{r}
y = a + b * x + rnorm(n = length(x), sd = 1)

(data = tibble(x, y))
```

```{r}
ggplot(data, aes(x = x, y = y)) +
  geom_point() +
  geom_abline(intercept = a, slope = b)
```

$$
y_i \sim \mathcal{N}(\mu_i, \sigma^2)
$$

$$
\begin{align}
\mu_i & = \beta_0 + \beta_1 x_i \\
& = \beta_0 \cdot 1_i + \beta_1 \cdot x_i \\
\end{align}
$$

R formula syntax:

$$
y \sim 1 + x
$$

```{r}
(design = model.matrix(y ~ 1 + x, data))
```

```{r}
fit = lm(y ~ 1 + x, data)
tidy(fit)
```

$$
\hat{y}_i = \hat{\beta_0} + \hat{\beta_1} x_i
$$

$$
e_i = y_i - \hat{y}_i
$$

```{r}
#| message: false
ggplot(data, aes(x = x, y = y)) +
  geom_point() +
  geom_abline(intercept = a, slope = b) +
  geom_abline(intercept = fit$coefficients[1], slope = fit$coefficients[2], color = "orange") # + geom_smooth(method = "lm", color = "orange")
```

### Regression model for covariates

A contrast is a linear combination of parameters/statistics. Another word for a linear combination is a weighted sum.

$$
\text{expression} = \beta_0 + \beta_1 \text{age}
$$

```{r}
set.seed(20190429)

mouse = paste0("MOUSE", 1:6)
age = c(1,2,3,4,5,6)
expression = 2 + 0.5 * age + rnorm(n = 6, sd = 0.1)

(data = tibble(mouse, expression, age))
```

```{r}
(design = model.matrix(expression ~ 1 + age, data))
```

```{r}
fit = lm(expression ~ 1 + age, data)
tidy(fit)
```

```{r}
model.name = paste0("E(y)=", round(fit$coef[1],2), "+", round(fit$coef[2],2), "*x")

ggplot(data, aes(y = expression, x = age)) +
  geom_point() +
  geom_abline(intercept = fit$coef[1],
              slope = fit$coef[2],
              color = "red") +
  labs(title = model.name,
       y = "expression (y)",
       x = "age (x)")
```

### Means model for factors

```{r}
genotype = fct_relevel(rep(c("WT", "KO"), each = 3), "WT")

(data = tibble(mouse, expression, genotype))
```

```{r}
(design = model.matrix(expression ~ 1 + genotype, data))
```

```{r}
fit = lm(expression ~ 1 + genotype, data)
tidy(fit)
```

```{r}
model.name = paste0("E(y)=", round(fit$coef[1],2), "+", round(fit$coef[2],2), "*KO")

ggplot(data, aes(y = expression, x = genotype)) +
  geom_point() +
  geom_hline(yintercept = fit$coef[1],
             color = "red") +
  geom_hline(yintercept = fit$coef[1] + fit$coef[2],
             color = "blue") +
  labs(title = model.name,
       y = "expression (y)",
       x = "genotype (x)")
```

### Mean-reference model for factors

```{r}
(design = model.matrix(expression ~ 0 + genotype, data))
```

```{r}
fit = lm(expression ~ 0 + genotype, data)
tidy(fit)
```

```{r}
model.name = paste0("E(y)=", round(fit$coef[1],2), "*WT", "+", round(fit$coef[2],2), "*KO")

ggplot(data, aes(y = expression, x = genotype)) +
  geom_point() +
  geom_hline(yintercept = fit$coef[1],
             color = "red") +
  geom_hline(yintercept = fit$coef[2],
             color = "blue") +
  labs(title = model.name,
       y = "expression (y)",
       x = "genotype (x)")
```

```{r}
avg_predictions(fit, df = insight::get_df(fit))
```

```{r}
(contrasts = makeContrasts(genotypeKO - genotypeWT, levels=colnames(design)))
```

```{r}
lmfit = lmFit(expression, design)

contrasts.fit(lmfit, contrasts)$coefficient
```

```{r}
avg_comparisons(fit, df = insight::get_df(fit))
```

## Design and contrast matrices

XXX

## Studies of treatments and control

### All pairwise comparisons

```{r}
treatment = as_factor(rep(c("CTRL", "A", "B", "C"), each=3))
expression = rep(c(0,1,2,4), each=3) + 1
expression = expression + rnorm(length(treatment), sd=0.1)
mouse = paste0("MOUSE", 1:length(treatment))

(data = tibble(mouse, expression, treatment))
```

```{r}
(design = model.matrix(expression ~ 1 + treatment, data))
```

```{r}
fit = lm(expression ~ 1 + treatment, data)
tidy(fit)
```

```{r}
avg_comparisons(fit, df = insight::get_df(fit))
```

```{r}
model.name = paste0("E(y)=", round(fit$coef[1],2), "+", round(fit$coef[2],2), "*A", "+", round(fit$coef[3],2), "*B", "+", round(fit$coef[4],2), "*C")

ggplot(data, aes(y = expression, x = treatment)) +
  geom_point() +
  geom_hline(yintercept = fit$coef[1],
             color = "red") +
  geom_hline(yintercept = fit$coef[1] + fit$coef[2],
             color = "blue") +
  geom_hline(yintercept = fit$coef[1] + fit$coef[3],
             color = "green") +
  geom_hline(yintercept = fit$coef[1] + fit$coef[4],
             color = "orange") +
  labs(title = model.name,
       y = "expression (y)",
       x = "treatment (x)")
```

```{r}
(design = model.matrix(expression ~ 0 + treatment, data))
```

```{r}
fit = lm(expression ~ 0 + treatment, data)
tidy(fit)
```

```{r}
avg_predictions(fit, by = "treatment", df = insight::get_df(fit))
```

```{r}
(contrasts = makeContrasts(
  treatmentA-treatmentCTRL,
  treatmentB-treatmentCTRL, 
  treatmentC-treatmentCTRL,
  levels=colnames(design)))
```

```{r}
lmfit = lmFit(expression, design)

contrasts.fit(lmfit, contrasts)$coefficient
```

```{r}
avg_comparisons(fit, df = insight::get_df(fit))
```

```{r}
model.name = paste0("E(y)=", round(fit$coef[1],2), "*CTRL", "+", round(fit$coef[2],2), "*A", "+", round(fit$coef[3],2), "*B", "+", round(fit$coef[4],2), "*C")

ggplot(data, aes(y = expression, x = treatment)) +
  geom_point() +
  geom_hline(yintercept = fit$coef[1],
             color = "red") +
  geom_hline(yintercept = fit$coef[2],
             color = "blue") +
  geom_hline(yintercept = fit$coef[3],
             color = "green") +
  geom_hline(yintercept = fit$coef[4],
             color = "orange") +
  labs(title = model.name,
       y = "expression (y)",
       x = "treatment (x)")
```

### Control vs the rest

```{r}
(contrasts = makeContrasts((treatmentA + treatmentB + treatmentC)/3 - treatmentCTRL, levels=colnames(design)))
```

```{r}
fit = lmFit(expression, design)

contrasts.fit(fit, contrasts)$coefficient
```

## Studies of interactions and additivity of treatments

### Interactions using a single factor model

Same as the multiple treatment models above but considering treatment C as the combination of treatment A and B and testing whether its effect (b3) equals the sum of the effects of A (b1) and B (b2).

```{r}
fit = lm(expression ~ 0 + treatment, data)
tidy(fit)
```

```{r}
avg_predictions(fit, by = "treatment", df = insight::get_df(fit))
```

```{r}
avg_comparisons(fit, df = insight::get_df(fit))
```

```{r}
avg_comparisons(fit, hypothesis = "b3 = b1 + b2", df = insight::get_df(fit))
```

```{r}
avg_comparisons(fit, hypothesis = "b3 - b1 - b2 = 0", df = insight::get_df(fit))
```

### Interactions using a two-factor model

```{r}
treat1 <- as_factor(c(0,0,0,1,1,1,0,0,0,1,1,1))
treat2 <- as_factor(c(0,0,0,0,0,0,1,1,1,1,1,1))
levels(treat1) = levels(treat2) = c("NO", "YES")

(data = tibble(mouse, expression, treat1, treat2))
```

```{r}
(design = model.matrix(expression ~ treat1 * treat2, data))
```

```{r}
fit = lm(expression ~ treat1 * treat2, data)
tidy(fit)
```

```{r}
avg_comparisons(fit, df = insight::get_df(fit))
```

### Additivity using a two-factor model

```{r}
(design = model.matrix(expression ~ treat1 + treat2, data))
```

```{r}
fit = lm(expression ~ treat1 + treat2, data)
tidy(fit)
```

```{r}
avg_comparisons(fit, df = insight::get_df(fit))
```

## Studies with multiple factors

### Conversion to a single factor

```{r}
tissue = treat1
levels(tissue) = c("LUNG", "BRAIN")
cells = treat2
levels(cells) = c("B", "T")
tissue_cells = paste(tissue, cells, sep="_")
tissue_cells = as_factor(tissue_cells)

(data = tibble(mouse, expression, tissue, cells, tissue_cells))
```

```{r}
(design = model.matrix(expression ~ 0 + tissue_cells, data))
```

```{r}
fit = lm(expression ~ 0 + tissue_cells, data)
tidy(fit)
```

```{r}
avg_predictions(fit, by = "tissue_cells", df = insight::get_df(fit))
```

### Accounting for factors that are not of interest

```{r}
treatment = as_factor(rep(LETTERS[1:4], each=3))
lane = rep(c("L1", "L2"), c(6,6))
lane = as_factor(sample(lane, length(treatment), replace=FALSE))
technician = c("I", "II")
technician = as_factor(sample(technician, length(treatment), replace=TRUE))

(data = tibble(mouse, expression, treatment, lane, technician))
```

```{r}
(design = model.matrix(expression ~ 0 + treatment + lane + technician, data))
```

```{r}
fit = lm(expression ~ 0 + treatment + lane + technician, data)
tidy(fit)
```

```{r}
avg_predictions(fit, by = "treatment", df = insight::get_df(fit))
```

```{r}
(contrasts = makeContrasts(treatmentB - treatmentA,
                           treatmentC - treatmentA,
                           treatmentD - treatmentA,
                           levels=colnames(design)))
```

```{r}
lmfit = lmFit(expression, design)

contrasts.fit(lmfit, contrasts)$coefficient
```

```{r}
avg_comparisons(fit, variables = c("treatment"), df = insight::get_df(fit))
```

## Studies with multiple covariates (and factors)

XXX

```{r}
sessionInfo()
```
